src/Util/Sniffs/Parentheses.php

```diff
@@ -0,0 +1,332 @@
+<?php
+/**
+ * Utility functions for use when examining nested parentheses.
+ *
+ * @author    Juliette Reinders Folmer <phpcs_nospam@adviesenzo.nl>
+ * @copyright 2019 Squiz Pty Ltd (ABN 77 084 670 600)
+ * @license   https://github.com/squizlabs/PHP_CodeSniffer/blob/master/licence.txt BSD Licence
+ */
+
+namespace PHP_CodeSniffer\Util\Sniffs;
+
+use PHP_CodeSniffer\Files\File;
+
+class Parentheses
+{
+
+
+    /**
+     * Get the pointer to the parentheses owner of an open/close parenthesis.
+     *
+     * @param \PHP_CodeSniffer\Files\File $phpcsFile The file where this token was found.
+     * @param int                         $stackPtr  The position of T_OPEN/CLOSE_PARENTHESIS token.
+     *
+     * @return int|false StackPtr to the parentheses owner or false if the parenthesis
+     *                   does not have a (direct) owner.
+     */
+    public static function getOwner(File $phpcsFile, $stackPtr)
+    {
+        $tokens = $phpcsFile->getTokens();
+        if (isset($tokens[$stackPtr], $tokens[$stackPtr]['parenthesis_owner']) === false) {
+            return false;
+        }
+
+        return $tokens[$stackPtr]['parenthesis_owner'];
+
+    }//end getOwner()
+
+
+    /**
+     * Check whether the parenthesis owner of an open/close parenthesis is within a
+     * limited set of valid owners.
+     *
+     * @param \PHP_CodeSniffer\Files\File $phpcsFile   The file where this token was found.
+     * @param int                         $stackPtr    The position of T_OPEN/CLOSE_PARENTHESIS token.
+     * @param int|string|array            $validOwners Array of token constants for the owners
+     *                                                 which should be considered valid.
+     *
+     * @return boolean True if the owner is within the list of $validOwners, false if not and
+     *                 if the parenthesis does not have a (direct) owner.
+     */
+    public static function isOwnerIn(File $phpcsFile, $stackPtr, $validOwners)
+    {
+        $owner = self::getOwner($phpcsFile, $stackPtr);
+        if ($owner === false) {
+            return false;
+        }
+
+        $tokens      = $phpcsFile->getTokens();
+        $validOwners = (array) $validOwners;
+        return in_array($tokens[$owner]['code'], $validOwners, true);
```

Comment:
I've set this class up largely to mirror what is done in the `Conditions/Scope` class, which also uses `in_array()` for `$types`.

I totally agree that using `isset()` would be faster, but it would also make the sniff more fiddly when passing parameters to it as a sniff writer, as you would always need to make sure that the `$validOwners` array would follow the `TOKEN_CONSTANT => true` pattern, which I think will lead to quite some bugs in sniffs written by less experienced sniff writers if/when a numerically indexed array with the token constants as values would be passed.

So in the end, I decided to go with the principle of least astonishment and let the `$validOwners` parameter be an ordinary array, just like `$types`.

I considered using an `array_flip()` on `$validOwners` to get round the overhead of `in_array()`, but that would probably have nearly as much overhead.

Reply:
Fair points.